File: src/carve/_accuracy.py
"""Accuracy-based generalizability metrics."""

import numpy as np

from ._utils import align_cluster_labels

# (sample_indices, true_labels, predicted_labels) triple from a prediction run
PredictionRun = tuple[np.ndarray, np.ndarray, np.ndarray]


def compute_generalizability_scores(
    n_samples: int,
    runs: list[PredictionRun],
) -> np.ndarray:
    """Compute per-sample generalizability from prediction runs.

    Parameters
    ----------
    n_samples : int
        Total number of samples in the original dataset.
    runs : list of tuple
        Each run is (sample_indices, true_labels, predicted_labels).

    Returns
    -------
    scores : ndarray of shape (n_samples,)
        Per-sample accuracy across available runs. Samples never evaluated
        receive score 0.
    """
    correct = np.zeros(n_samples, int)
    total = np.zeros(n_samples, int)

    for sample_idx, true_labels, predicted_labels in runs:
        aligned_predictions = align_cluster_labels(true_labels, predicted_labels)
        correct_mask = (true_labels == aligned_predictions).astype(int)
        np.add.at(correct, sample_idx, correct_mask)
        np.add.at(total, sample_idx, 1)

    out = np.zeros(n_samples, float)
    valid = total > 0
    out[valid] = correct[valid] / total[valid]
    return out

File: carve-r/R/accuracy.R
# Accuracy-based generalizability scoring. Mirrors _accuracy.py.


#' Per-sample generalizability scores
#'
#' Mirrors `_accuracy.compute_generalizability_scores`. For each sample,
#' accumulates prediction-correct counts across runs (after Hungarian
#' alignment to the true labels) and returns `correct / total`. Samples
#' that were never evaluated receive `0`.
#'
#' @param n_samples Integer; total number of samples in the original dataset.
#' @param runs List of runs. Each run is a list with components
#'   `sample_idx` (1-based integer indices), `labels` (true labels), and
#'   `predicted` (predicted labels).
#'
#' @return Numeric vector of length `n_samples`.
#'
#' @export
compute_generalizability_scores <- function(n_samples, runs) {
  correct <- integer(n_samples)
  total <- integer(n_samples)

  for (r in runs) {
    aligned <- align_cluster_labels(r$labels, r$predicted)
    correct_mask <- as.integer(r$labels == aligned)
    idx <- r$sample_idx
    correct[idx] <- correct[idx] + correct_mask
    total[idx] <- total[idx] + 1L
  }

  out <- numeric(n_samples)
  valid <- total > 0L
  out[valid] <- correct[valid] / total[valid]
  out
}
